[MRG + 1] check that unfitted estimators raise ValueError

TomDLT opened the pull request:
> Issue : check_estimator requires raising NotFittedError #4661

Please tell me what you think

TomDLT committed:
ENH check that unfitted estimators raise ValueError

agramfort commented:
LGTM

@amueller ?

amueller commented:
The reason you needed to add a custom function was that we are fine with either ``AttributeError`` or ``ValueError``, right? Maybe just add a variant of ``assert_raise_message`` that accepts a list of allowed Errors?

TomDLT commented:
Right, we want to accept both `AttributeError` or `ValueError`.
I can add a variant of `assert_raise_message`, but then I don't see how to add a message suggesting the use of `NotFittedError`. Do you have an idea? 
Do we want to drop the suggestion to `NotFittedError`?

amueller commented:
Yeah, I don't think it is necessary.

ogrisel commented on sklearn/utils/estimator_checks.py:
```diff
@@ -787,19 +787,48 @@ def check_estimators_unfitted(name, Estimator):
+
+        *args : the positional arguments to `func`.
+
+        **kw : the keyword arguments to `func`
```
Docstring indentation issue: the Parameters block need to be at the same level as the opening `"""`.

ogrisel commented on sklearn/utils/estimator_checks.py:
```diff
@@ -787,19 +787,48 @@ def check_estimators_unfitted(name, Estimator):
+
+def check_unfitted(name, func, *args, **kwargs):
+    """Helper function to check that an AttributeError or a ValueError
+    is raised when calling 'func' on unfitted estimator
```
Please break that line and give more details:

```python
def check_unfitted(name, func, *args, **kwargs):
    """Check that func raises an exception indicating an unfitted estimator

    Unfitted estimators should raise either AttributeError or ValueError. The
    specific exception type NotFittedError inherits from both and can therefore
    be adequately raised  for that purpose.

    Parameters
    ---------------
    ... 
    """

ogrisel commented on sklearn/utils/estimator_checks.py:
```diff
@@ -787,19 +787,48 @@ def check_estimators_unfitted(name, Estimator):
+    except (ValueError, AttributeError):  # correct Error
+        pass
+    except:
+        correct_raise = False
```
This could be simplified as follows:

```python
correct_raise = False
try:
    func(*args, **kwargs)
except (ValueError, AttributeError):  # correct Error
    correct_raise = True

if not correct_raise:
    ...
```

ogrisel replied:
If an unexpected exception type is raised, I think it should not be caught to make it easier to debug the issue.

ogrisel commented on sklearn/utils/tests/test_estimator_checks.py:
```diff
@@ -6,10 +6,17 @@
 
+class CorrectNotFittedError(ValueError):
+    """Exception class to raise if estimator is used before fitting.
+    Like NotFittedError, it inherits from ValueError, but not from
```
Please add a blank line before this line to follow PEP 257.

TomDLT commented:
I reverted and deleted my new function.
Instead, I use `assert_raises`, which actually accepts multiple errors in a tuple.

We loose the detailed message, but we avoid to create a very specific function.

amueller commented:
I think it would be nice to use ``assert_raise_regex``, though, to check for a helpful message. Currently the check is very loose.

TomDLT commented:
New try: I extended `assert_raise_message` in order to handle tuple of exception 
(`assert_raises` handles tuples of exception, but `assert_raises_regex` does not)

The current test checks if the message contains `fit`, as the messages for unfitted estimators can be very different. Do you want me to change all the message, for example by using more systematically `check_is_fitted` ?

amueller commented:
Hum... I'm not a particular fan of the escaping behavior in assert_raise_message... but I guess it is the right place to put this. So +1 for the approach. I guess checking for "fit" is fine. I haven't looked at the message.

amueller commented on sklearn/utils/testing.py:
```diff
@@ -387,10 +388,46 @@ def _assert_allclose(actual, desired, rtol=1e-7, atol=0,
+                                 "%r. %r does not." %
+                                 (message, error_message))
+
+    if not_raised:
```
you can just put this into the "else" of ``try: ... except: ... else: ...``

TomDLT replied:
Sure

amueller commented on sklearn/utils/testing.py:
```diff
@@ -387,10 +388,46 @@ def _assert_allclose(actual, desired, rtol=1e-7, atol=0,
+    if not_raised:
+        # concatenate exception names
+        if isinstance(exceptions, tuple):
+            names = ""
```
``names = " or ".join(e.__name__ for e in exceptions)`` maybe?

TomDLT replied:
Excellent, I knew it should exist a better way

amueller commented:
LGTM

ogrisel commented on sklearn/utils/testing.py:
```diff
@@ -387,10 +388,40 @@ def _assert_allclose(actual, desired, rtol=1e-7, atol=0,
+    *args : the positional arguments to `func`.
+
+    **kw : the keyword arguments to `func`
+   """
```
bad indentation

ogrisel commented:
That's a good opportunity to get rid of the escaping behavior of `assert_raise_message`. The following seems to work:

```
diff --git a/sklearn/utils/testing.py b/sklearn/utils/testing.py
index cc14b3a..7f6039c 100644
--- a/sklearn/utils/testing.py
+++ b/sklearn/utils/testing.py
@@ -402,16 +402,14 @@ def assert_raise_message(exceptions, message, function, *args, **kwargs):
     *args : the positional arguments to `func`.
 
     **kw : the keyword arguments to `func`
-   """
-    message = re.escape(message)
-
+    """
     try:
         function(*args, **kwargs)
     except exceptions as e:
         error_message = str(e)
-        if not re.compile(message).search(error_message):
-            raise AssertionError("Error message should match pattern "
-                                 "%r. %r does not." %
+        if message not in error_message:
+            raise AssertionError("Error message does not include the expected"
+                                 " string: %r. Observed error message: %r" %
                                  (message, error_message))
     else:
         # concatenate exception names
@@ -543,7 +541,7 @@ DONT_TEST = ['SparseCoder', 'EllipticEnvelope', 'DictVectorizer',
 def all_estimators(include_meta_estimators=False,
                    include_other=False, type_filter=None,
                    include_dont_test=False):
-                   
+
     """Get a list of all estimators from sklearn.
 
     This function crawls the module and gets all classes that inherit
```

I fixed the indentation issue in that patch.

Other than that +1 as well.

ogrisel commented:
@TomDLT please also squash the commits and rebase on master.

ogrisel commented on sklearn/utils/tests/test_estimator_checks.py:
```diff
@@ -6,10 +6,18 @@
+    """Exception class to raise if estimator is used before fitting.
+
+    Like NotFittedError, it inherits from ValueError, but not from
+    AttributeError. Used for testing only"""
```
cosmetics: closing `"""` should be on their own line.

ogrisel commented on sklearn/utils/testing.py:
```diff
@@ -508,11 +537,10 @@ def uninstall_mldata_mock():
                    include_other=False, type_filter=None,
                    include_dont_test=False):
-                   
+
```
cosmetics: no need for a blank line here. It can be removed entirely for consistency.

ogrisel commented on sklearn/utils/estimator_checks.py:
```diff
@@ -796,8 +796,11 @@ def check_estimators_fit_returns_self(name, Estimator):
+
+    Unfitted estimators should raise either AttributeError or ValueError.
+    The specific exception type NotFittedError inherits from both and can
+    therefore be adequately raised for that purpose."""
```
cosmetics: closing `"""` on its own line (PEP257).

TomDLT commented:
Cosmetic changes done!

amueller commented:
Thanks @TomDLT  :)
